sinks/gcm/driver.go

```diff
@@ -81,69 +83,85 @@ type metricDescriptor struct {
-			description: "Rate of change of time since start in seconds per second",
-		},
-		{
-			name:        "cpu/usage",
-			description: "Rate of total CPU usage in millicores per second",
-		},
-		{
-			name:        "memory/page_faults",
-			description: "Rate of major page faults in counts per second",
-		},
-		{
-			name:        "network/rx",
-			description: "Rate of bytes received over the network in bytes per second",
-		},
-		{
-			name:        "network/rx_errors",
-			description: "Rate of errors sending over the network in errors per second",
-		},
-		{
-			name:        "network/tx",
-			description: "Rate of bytes transmitted over the network in bytes per second",
-		},
-		{
-			name:        "network/tx_errors",
-			description: "Rate of errors transmitting over the network in errors per second",
-		},
-	}
-
-	// Replac the description if we have an alternate one.
-	for _, desc := range descriptions {
-		if metric.Name == desc.name {
-			return desc.description
-		}
-	}
-
-	return metric.Description
-}
-
 // Map of metric name to translation function.
 var translationFuncs = map[string]func(float64) float64{
-	"uptime": func(value float64) float64 {
+	"uptime_rate": func(value float64) float64 {
 		// Convert from milliseconds to seconds.
 		return value / 1000
 	},
-	"cpu/usage": func(value float64) float64 {
+	"cpu/usage_rate": func(value float64) float64 {
 		// Convert from billionths of a core to millicores.
 		return value / 1000000
 	},
 }
 
+type listMetricsResponse struct {
+	Metrics []metricDescriptor `json:"metrics,omitempty"`
+}
+
+func (self *gcmSink) defaultUrlPath() string {
+	return fmt.Sprintf("https://www.googleapis.com/cloudmonitoring/v2beta2/projects/%s", self.project)
+}
+
+func (self *gcmSink) listMetrics() error {
+	response := listMetricsResponse{}
+	url, err := url.Parse(self.defaultUrlPath() + "/metricDescriptors")
+	if err != nil {
+		return err
+	}
+	err = sendRequest("GET", self.token, url, nil, &response)
+	if err != nil {
+		glog.Errorf("[GCM] list metrics failed %v", err)
+		return err
+	}
+	apiPrefix := fmt.Sprintf("%s/%s", customApiPrefix, metricDomain)
+	for idx, m := range response.Metrics {
+		if strings.HasPrefix(m.Name, apiPrefix) {
+			self.exportedMetrics[m.Name] = response.Metrics[idx]
+		}
+	}
+
+	return nil
+}
+
+func (self *gcmSink) deleteMetric(metricName string) error {
+	url := &url.URL{
+		Scheme: "https",
+		Host:   "www.googleapis.com",
+		Opaque: fmt.Sprintf("%s/metricDescriptors/%s", self.defaultUrlPath(), url.QueryEscape(metricName)),
+	}
+
+	err := sendRequest("DELETE", self.token, url, nil, nil)
+	if err != nil {
+		glog.V(2).Infof("[GCM] Deleting metric %q failed: %v", metricName, err)
+	}
+	return err
+}
+
+func (self *gcmSink) addMetric(request metricDescriptor) error {
+	if err := self.listMetrics(); err != nil {
+		return err
+	}
+	if existingMetric, found := self.exportedMetrics[request.Name]; found {
+		if existingMetric.TypeDescriptor != request.TypeDescriptor ||
```

Comment:
You can check for only type, if the description is different the add below would update it.